### ISSUE TITLE
Investigate the cause of the progress stuck at 99% in Production

### ISSUE THREAD

**Issue by diversemix** [elifesciences/elife-xpub#1491]
See #1487 

There was a sequence of event that resulted in an error on the server (ScienceBeam?) that resulted in the state that the user saw of the manuscript upload being stuck at 99%.

Use Loggly to investigate and raise issues accordingly

**Comment by giorgiosironi** [comment]
[Exhausting memory](https://infrastructure.newrelic.com/accounts/1451451/hosts?filters=%7B%22and%22%3A%5B%7B%22is%22%3A%7B%22entityName%22%3A%22elife-xpub--prod--1%22%7D%7D%5D%7D&timeEnd=1549710420000&timeStart=1549587600000) is likely the cause:
![f59f5d6d-0874-4c3f-bbd3-0c954dd5089f](https://user-images.githubusercontent.com/160299/52559582-84cc6e80-2ded-11e9-863d-8915349991b7.png)
But we don't know yet what filled the memory, though we have suspects.

**Comment by giorgiosironi** [comment]
It was a Java process ([in yellow](https://infrastructure.newrelic.com/accounts/1451451/processes?filters=%7B%22and%22%3A%5B%7B%22or%22%3A%5B%7B%22like%22%3A%7B%22apmApplicationNames%22%3A%22%7Celife-xpub--prod%7C%22%7D%7D%5D%7D%5D%7D&scope=New%20filter%20set&timeEnd=1549883580000&timeFocused=1549645980000&timeStart=1549587600000), black is just the average of all processes):
![screenshot_2019-02-11 infrastructure by new relic](https://user-images.githubusercontent.com/160299/52559723-fefcf300-2ded-11e9-9961-8d488601e46c.png)
Hence it should be Grobid.

**Comment by giorgiosironi** [comment]
It seems [both hosts were potentially affected](https://infrastructure.newrelic.com/accounts/1451451/hosts?filters=%7B%22and%22%3A%5B%7B%22or%22%3A%5B%7B%22like%22%3A%7B%22apmApplicationNames%22%3A%22%7Celife-xpub--prod%7C%22%7D%7D%5D%7D%5D%7D&timeEnd=1549883580000&timeFocused=1549645980000&timeStart=1549587600000&scope=New%20filter%20set):
![screenshot_2019-02-11 infrastructure by new relic 1](https://user-images.githubusercontent.com/160299/52559915-af6af700-2dee-11e9-93df-f7393bf1bafe.png)

**Comment by diversemix** [comment]
This is likely to be grobid ... @de-code - would this be your diagnosis too?
Restarting this on a regular basis - might be part of #1436

**Comment by de-code** [comment]
Yes, GROBID will likely be the Java process. Did it get OOMed?

Not sure whether we talked about it before, does the machine maybe not have enough memory?

**Comment by Cooryd** [comment]
From close #1513 
---
**Describe the bug**

See #1487 for full context.

It has been reported that a manuscript upload can become 'stuck' at 99%. This should be investigated to see if there are any improvements that can be made to prevent this from happening.

**Steps To Reproduce**

This is the part that needs investigating.

As a starting point it may be worth looking for any scenarios where the `setInterval` that is used to relay predicted progress up until 99% continues to run after an error is encountered.

**Expected behavior**

The progress status view of the upload in the client should only ever be shown when there is something actively uploading. Any errors should be handled correctly and displayed as appropriate and the progress 'messages' should stop.

**Comment by giorgiosironi** [comment]
Yesterday we had a similar alert on memory usage reaching 85% (and the host ceasing to report to New Relic):
https://infrastructure.newrelic.com/accounts/1451451/hosts?filters=%7B%22and%22%3A%5B%7B%22is%22%3A%7B%22entityName%22%3A%22elife-xpub--prod--2%22%7D%7D%5D%7D&timeStart=6%20hours%20ago

**Comment by Cooryd** [comment]
I would take a guess that https://github.com/elifesciences/elife-xpub/blob/97818f90cbf3132fa4bb70472778528daed5026d/server/xpub-controller/manuscript.js#L117 is the point of execution that threw the error if it relates to science beam being the eater of memory. We're going to try to better deal with the client impact of this issue in #1521 but this won't resolve the root cause of the error thrown

**Comment by diversemix** [comment]
@jyenes ... can you comment on why this was re-opened?

**Comment by jyenes** [comment]
It was in TODO column and I closed by mistake, so I left it in the same status that it was before

**Comment by hem-brahmbhatt** [comment]
The discussion here leads me to believe that the "out of memory" issue is the error we're trying to reproduce.

My approach would be to run a load test on staging to see if it can be reproduced.

**Comment by diversemix** [comment]
The last cause of the progress bar getting stuck at 99% is the fact that the predicted time is in certain cases incorrect. I've looked at the logs and analysed the experience users have had given the prediction being wrong.

- 16% of users get stuck at 99% for an average of 34 seconds. (worst case was 941 seconds)
- 84% of users have their progress finished before 100% ( worst case is 407 seconds)

A graph of absolute errors is shown/attached (where y-axis is in seconds, -ve seconds indicate progress was finished before 100%)

**Comment by diversemix** [comment]
![image](https://user-images.githubusercontent.com/8524685/55016413-986f1580-4fe6-11e9-8792-5a9085e6325a.png)

**Comment by diversemix** [comment]
Second analysis was done on % error... this shows the following result. Which indicates that if the prediction is doubled then this would include all outliers. @hdrury1 this maybe a good "quick-fix"
...

**Comment by diversemix** [comment]
![image](https://user-images.githubusercontent.com/8524685/55016548-e71caf80-4fe6-11e9-901b-ff188ea6abbf.png)

**Comment by diversemix** [comment]
@hdrury1 / @chugginselifesciences ... marking this as closed - happy though to carry chatting on here if there is anything unclear.

**Comment by chris-huggins** [comment]
@diversemix Just to check I understand, does doubling the estimate mean that we'll see either/both:

- An increase in the number or % of people seeing their upload progress 'jump' to 100% (assuming that's what currently happens to the `84% of users have their progress finished before 100%`) 

- The difference between the estimate and actual progress will increase for those users that have their progress finish before 100%? In other words, for those that the progress jumps to 100%, will they now see their progress fed back to them slower initially, therefore jumping to 100% earlier (e.g. from 30% to 100% instead of 60%)?

So, to summarise, am I right in thinking the doubling reduces the likelihood of serious problems, but means uploads appear slower to users, which means they 'jump' to 100% earlier?

**Comment by diversemix** [comment]
@chugginselifesciences / @hdrury1 ... I've re-analysed the results this morning - because I suspected something I did wasn't right.... and it wasn't - there is no "quick fix" like doubling the prediction. We are going to have to adapt the prediction.

**Comment by diversemix** [comment]
Data used for analysis...
[upload-predictions.zip](https://github.com/elifesciences/elife-xpub/files/3012302/upload-predictions.zip)

**Comment by de-code** [comment]
I've just checked the logs and it seems xPub isn't making use of the `include` parameter. I raised #1773
That could maybe keep the conversion time down. Otherwise I guess there will be a correlation with file size?

**Comment by chris-huggins** [comment]
To bring my question from yesterday into Github, I just want to check we're not overcomplicating things with this prediction... would it be possible to instead feedback 'actual' file upload progress instead of a 'predicted progress'? This would involve removing the sciencebeam progress from the percentage feedback but we could show it afterwards with general progress indicator.

But @diversemix am I right in understanding that this isn't possible? I feel like if it was possible this would solve a lot of our problems caused by inaccurate predictions.

If it were possible I see it working along these lines: the drop target would show spinning feedback, plus file upload progress expressed as a percentage, followed by something like "Extracting file" (and the spinner continuing) while sciencebeam attempts extraction for maximum 10 seconds before timing out and continuing without a title. Only once sciencebeam finishes extraction or times out do we show "uploaded" state on the drop target
